help_function/src/onedpl_test_tagged_pointer.cpp

```diff
@@ -0,0 +1,153 @@
+                          : "dpct::device_sys_tag";
+  std::string int_ptr_name = "dpct::tagged_pointer<" + sys + ", int>";
+  std::string int_wrapper_name =
+      "dpct::tagged_pointer<" + sys + ", integer_wrapper>";
+  std::string void_ptr_name = "dpct::tagged_pointer<" + sys + ", void>";
+
+  dpct::tagged_pointer<SystemTag, void> void_ptr_beg =
+      dpct::malloc(system, sizeof(int) * n);
+  dpct::tagged_pointer<SystemTag, int> int_ptr_beg =
+      static_cast<dpct::tagged_pointer<SystemTag, int>>(void_ptr_beg);
+
+  dpct::tagged_pointer<SystemTag, void> void_ptr_beg2 =
+      static_cast<dpct::tagged_pointer<SystemTag, void>>(int_ptr_beg);
+  failing_tests += ASSERT_EQUAL(void_ptr_name + " conversion operator",
+                                void_ptr_beg == void_ptr_beg2, true);
+
+  dpct::tagged_pointer<SystemTag, int> int_ptr_end = int_ptr_beg + n;
+  failing_tests += ASSERT_EQUAL(
+      int_ptr_name + " add operator",
+      static_cast<int *>(int_ptr_end) - static_cast<int *>(int_ptr_beg), n);
+
+  dpct::tagged_pointer<SystemTag, int> expect_beg = int_ptr_end - n;
+  failing_tests += ASSERT_EQUAL(int_ptr_name + " subtract operator",
+                                int_ptr_beg == expect_beg, true);
+
+  failing_tests += ASSERT_EQUAL(int_ptr_name + " difference operator",
+                                int_ptr_end - int_ptr_beg, n);
+
+  expect_beg++;
+  expect_beg--;
+  failing_tests +=
+      ASSERT_EQUAL(int_ptr_name + " postfix increment and decrement",
+                   int_ptr_beg == expect_beg, true);
+
+  ++expect_beg;
+  --expect_beg;
+  failing_tests +=
+      ASSERT_EQUAL(int_ptr_name + " prefix increment and decrement",
+                   int_ptr_beg == expect_beg, true);
+
+  expect_beg += 2;
+  expect_beg -= 2;
+  failing_tests +=
+      ASSERT_EQUAL(int_ptr_name + " addition and subtraction assignment",
+                   int_ptr_beg == expect_beg, true);
+
+  // Test conversion to base pointer
+  int *int_ptr_beg_raw = int_ptr_beg;
+  int *int_ptr_end_raw = int_ptr_end;
+  failing_tests += ASSERT_EQUAL(int_ptr_name + " conversion to int*",
+                                int_ptr_end_raw - int_ptr_beg_raw, n);
+
+  // device allocations use malloc_shared so this is safe
+  *int_ptr_beg = 4;
+  failing_tests += ASSERT_EQUAL(int_ptr_name + " dereference operator",
+                                *int_ptr_beg == 4, true);
+  int_ptr_beg[1] = 2;
+  failing_tests += ASSERT_EQUAL(int_ptr_name + " subscript operator",
+                                int_ptr_beg[1] == 2, true);
+
+  failing_tests += ASSERT_EQUAL(int_ptr_name + " subscript operator",
+                                int_ptr_beg[1] == 2, true);
+
+  dpct::tagged_pointer<SystemTag, integer_wrapper> int_wrapper_beg =
+      dpct::malloc<integer_wrapper>(system, 1);
+  int_wrapper_beg->val = 5;
+  failing_tests += ASSERT_EQUAL(int_wrapper_name + " arrow operator",
+                                int_wrapper_beg->val == 5, true);
```

Comment:
Perhaps check this with `(*int_wrapper_beg).val == 5`.

Reply:
I have made this change.